### PR TITLE
CHANGE: Re-enable temporarily disabled tests due to ProjectWide actions (ISX-1455)

### PR DESCRIPTION
**Pull request by jamesmcgill**
### Description

When Project wide actions landed a number of tests were disabled due to last minute bugs being found.
This PR fixes those issues and re-enables all tests.
Some tests still have project wide actions disabled. Each of those have been reviewed and accepted as having a genuine reason to disable the actions for the test. An explanation comment has been provided to each one in this PR too.

### Changes made

- Fix for the nasty binding resolution assert with a test.
- Document reasons on tests where project wide actions are disabled
- Stop using EditorBuildSettings as it was unnecessary. AssetDatabase is sufficient.

### Checklist

Before review:

- [ ] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [ ] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [ ] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

### CODE REVIEW

**Comment by jfreire-unity** [comment]
Could you rebase this from `develop`? Seems there are some commits that were already merged to `develop`. So I don't know from which commit should I be reviewing this.

File: Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs [hunk 1]
```diff
@@ -19,14 +19,12 @@ public void OnPreprocessBuild(BuildReport report)
             var preloadedAssets = PlayerSettings.GetPreloadedAssets();
 
 #if UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS
-            EditorBuildSettings.TryGetConfigObject(InputSettingsProvider.kEditorBuildSettingsActionsConfigKey,
-                out m_ProjectWideActions);
-
+            m_ProjectWideActions = Editor.ProjectWideActionsAsset.GetOrCreate();
```
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs:22]: Seems reasonable and was feedback in a previous PR.
I'm note sure why the config key was used previously though. Assuming no side effects this looks tidier.

File: ProjectSettings/ProjectSettings.asset [hunk 2]
```diff
@@ -120,8 +120,6 @@ PlayerSettings:
     Others: 1
   bundleVersion: 1.0
   preloadedAssets:
-  - {fileID: 0}
```
Comment (lyndon-unity) [inline ProjectSettings/ProjectSettings.asset:123]: unclear on the impact of this part but looks like a sensible removal

File: Assets/Tests/InputSystem/CoreTests_Actions.cs [hunk 3]
```diff
@@ -529,7 +529,7 @@ public void Actions_CanTargetSameControlWithMultipleActions()
     {
 #if UNITY_INPUT_SYSTEM_PROJECT_WIDE_ACTIONS
         // Exclude project-wide actions from this test
-        InputSystem.actions?.Disable();
+        InputSystem.actions?.Disable(); // Prevent these actions appearing in the `InputActionTrace`
```
Comment (lyndon-unity) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs:532]: Great to see the extra comments explaining why disabled.
I did wonder if we need some re-enabling?

File: Assets/Tests/InputSystem/CoreTests_Actions.cs [hunk 4]
```diff
@@ -1337,6 +1337,59 @@ public unsafe void Actions_CanTriggerBindingResolutionOnAction_FromCallback(bool
...
+        public const float EXPECTED_VALUE = 0.5f;
+        public override float Process(float value, InputControl control)
+        {
+            return EXPECTED_VALUE;
+        }
+    }
+
+    [Test]
+    [Category("Actions")]
+    public void Actions_ActiveBindingsHaveCorrectBindingIndicesAfterBindingResolution()
```
Comment (lyndon-unity) [inline Assets/Tests/InputSystem/CoreTests_Actions.cs:1352]: Nice to see a new test but was a little surprised as it doesn't seem related to re-enabling tests.

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs [hunk 5]
```diff
@@ -551,6 +551,7 @@ private void RestoreActionStatesAfterReResolvingBindings(UnmanagedMemory oldStat
                 newActionState.pressedInUpdate = oldActionState.pressedInUpdate;
                 newActionState.releasedInUpdate = oldActionState.releasedInUpdate;
                 newActionState.startTime = oldActionState.startTime;
+                newActionState.bindingIndex = oldActionState.bindingIndex;
```
Comment (lyndon-unity) [inline Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs:554]: Would this have any problems for the existing 1.8.0-pre.1 release ?
Do we need to hold back the release for this fix?

**Review by lyndon-unity — APPROVED** [review]
Looks like a reasonable change

File: Assets/Tests/InputSystem/CoreTests_Actions.cs (hunk 4 above)
Comment (lyndon-unity) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs:1352]: Ah I see it in the description - thanks

File: Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs (hunk 5 above)
Comment (lyndon-unity) [reply Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs:554]: Also is there a related bug regarding the assert, I can't find it ATM

Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/Actions/InputActionState.cs:554]: This PR is against ISX-1455. This is one of the bugs that prevented all the tests being enabled after project-wide actions was added.

File: Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs (hunk 1 above)
Comment (jamesmcgill) [reply Packages/com.unity.inputsystem/InputSystem/Editor/Settings/InputSettingsBuildProvider.cs:22]: I added it recently based on the documentation for using preloaded assets, but on reflection it is unnecessary in our case as we still have access to `AssetDatabase` when this code is run.

File: ProjectSettings/ProjectSettings.asset (hunk 2 above)
Comment (jamesmcgill) [reply ProjectSettings/ProjectSettings.asset:123]: It's an null preloaded asset which is just noise and only affects our project (not user projects)

File: Assets/Tests/InputSystem/CoreTests_Actions.cs (hunk 3 above)
Comment (jamesmcgill) [reply Assets/Tests/InputSystem/CoreTests_Actions.cs:532]: All tests are enabled in this PR.
All tests have project-wide actions enabled where appropriate.
Everywhere project-wide actions is disabled has been reviewed and a note added for the rationale.

**Comment by jamesmcgill** [comment]
@jfreire-unity I've rebased now.

**Review by jfreire-unity — APPROVED** [review]
